pkg/annotation/annotation.go

```diff
@@ -0,0 +1,199 @@
+	//    for example:
+	//       //+gocover:ignore:5
+	//       pf, err := os.Open(fileName)  -|
+	//       if err != nil {                |
+	// 	         return nil, err            | -> these 5 lines will be ignored
+	//       }                              |
+	//       defer pf.Close()              -|
+	//
+	//       profile, err := parseIgnoreProfilesFromReader(pf) -|
+	//       profile.Filename = fileName                        | -> code block
+	//       return profile, err                               -|
+	//
+	// Question: how to ignore the whole go function?
+	IgnoreRegexp = regexp.MustCompile(`^\s*//\s*\+gocover:ignore:([0-9A-Za-z]+)`)
+)
+
+// IgnoreType indicates the type of the ignore profile.
+// - ALL_IGNORE means the profile ignore the whole input file.
+// - BLOCK_IGNORE means the profile ignore several code block of the input file.
+type IgnoreType string
+
+const (
+	ALL_IGNORE   IgnoreType = "all"
+	BLOCK_IGNORE IgnoreType = "block"
+)
+
+// IgnoreProfile represents the ignore profiling data for a specific file.
+type IgnoreProfile struct {
+	// type of the ignore profile.
+	// when it's BLOCK_IGNORE, Lines includes all the code line number that should be ignored,
+	// IgnoreBlocks contains the concrete ignore data.
+	Type         IgnoreType
+	Filename     string
+	Lines        map[int]bool
+	IgnoreBlocks []*IgnoreBlock
+}
+
+// IgnoreBlock represents a single block of ignore profiling data.
+type IgnoreBlock struct {
+	Annotation string   // concrete ignore pattern
+	Contents   []string // ignore contents
+	Lines      []int    // corresponding code line number of the ignore contents
+}
+
+// ParseIgnoreProfiles parses ignore profile data in the specified file and returns a ignore profile.
+func ParseIgnoreProfiles(fileName string) (*IgnoreProfile, error) {
+	pf, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer pf.Close()
+
+	profile, err := parseIgnoreProfilesFromReader(pf)
+	profile.Filename = fileName
+	return profile, err
+}
+
+// parseIgnoreProfilesFromReader parses ignore profile data from the Reader and returns a ignore profile.
+func parseIgnoreProfilesFromReader(rd io.Reader) (*IgnoreProfile, error) {
+	s := bufio.NewScanner(rd)
+	lineNo := 0
+
+	profile := &IgnoreProfile{
+		Lines: make(map[int]bool),
+		Type:  BLOCK_IGNORE,
+	}
+
+	for s.Scan() {
+		lineNo++
+		line := s.Text()
+		match := IgnoreRegexp.FindStringSubmatch(line)
+
+		// not match, continue next line
+		if match == nil {
+			continue
+		}
+		if len(match) < 2 {
+			continue
+		}
+
+		// match contains the result of the regexp on IgnoreRegexp
+		// match = ["//+gocover:ignore:all", "all"] when input is `//+gocover:ignore:all`,
+		// match = ["//+gocover:ignore:block", "block"] when input is `//+gocover:ignore:block`,
+		// match = ["//+gocover:ignore:10", "10"] when input is `//+gocover:ignore:10`,
+		ignoreKind := match[1]
+		if ignoreKind == "all" { // set type to ALL_IGNORE and skip further processing
```

Comment:
similar question in L50 for the `IgnoreRegexp`, if we are interesting only about the `all` / `block` / number, can we just simplify the regex?